PySCF Converter

anbenali opened the pull request:
Issue remaining:
To this point, we only support basis set up to the G orbital, either with spherical or cartesian representation. Historically we have been following the Gamess Ordering of the AO. that order is not necessarily the best and to my knowledge no other code uses it. NWCHem and PYSCF use the same order and can handle higher order orbiltals. Maybe it is time to change th way we do it. The internal order in QMCPACK is located here: 
qmcpack/src/Numerics/CartesianTensor.h





***
Implemented: 
- Pyscf Converter file in Python for Molecules AND Solids (real and Complex WF)
- Convert4qmc -pyscf System.h5: For Molecules only. 
*****
Status:
- Molecules fully implemented and tested with spherical and cartesian orbitals. 
- Solids WF with Kpoints is stored properly. 
***
How to test/use.

Add the path to the PyscfToQmcpack.py to your PYTHONPATH.
At the end of your pyscf input add:

```
Title='LiH'

from PyscfToQmcpack import savetoqmcpack
savetoqmcpack(cell,mf,title)
```
If Solids:
```
Title='LiH'

from PyscfToQmcpack import savetoqmcpack
savetoqmcpack(cell,mf,title,kpts)
```

If title is left out, title="default"

in convet4qmc a tag is added:
```
convert4qmc -pyscf title.h5
```

This will generate the wfs, input and ptcl files.

prckent commented:
For an initial molecular test and example, I suggest recreating the GAMESS LiH molecule all electron tests from tests/molecules/LiH_dimer_ae/gamess.  The mean field wavefunction and energy should be exactly the same from pyscf, other than harmless normalization, basis function ordering, and numerical convergence differences.

prckent commented:
OK to test

ye-luo commented:
The last commit adds over 300 files which is not acceptable. Please use an archive or minimal input files which can reproduce the run.

anbenali commented:
Tared the 240 files for the original quantum package format.

anbenali commented:
Fixed

On Wed, Jan 3, 2018 at 10:24 PM, Ye Luo <notifications@github.com> wrote:

> The last commit adds over 300 files which is not acceptable. Please use an
> archive or minimal input files which can reproduce the run.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/QMCPACK/qmcpack/pull/607#issuecomment-355196076>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AXl6w2jVVMIdy-juYMrfHqBhf5qaZQ1rks5tHFJ-gaJpZM4RKg3E>
> .
>



-- 
-----------------------
Anouar Benali, PhD
Leadership Computing Facility
Argonne National Laboratory
Building 240 Office - 2127
9700 S Cass Av., Argonne Il, 60439
(630) 252-0058

ye-luo commented:
Does name="Gaussian" affect code path or it is just a name?
We don't need to test all the three 16-1, 4-4, 1-16 MPI-OMP combinations.
If I use hdf5, do I have identical qmcpack input files for gamess, QP, pyscf?

anbenali commented:
>>Does name="Gaussian" affect code path or it is just a name?

The name doesn't affect the run. However it is useful to know which
basisset was used. Pyscf provides the name and so will quantum Package. Not
sure about gamess. Gaussian will be the default name if none is found.

>>We don't need to test all the three 16-1, 4-4, 1-16 MPI-OMP combinations.
Great! I followed the previous practice. However it might be useful to have
4 4. I discovered a bug (variable not bcasted) with the 4 4.

>> If I use hdf5, do I have identical qmcpack input files for gamess, QP,
pyscf?

Yes. Pyscf h5 file contains more information necessary to generate the
particle sets and the WF, but once you have qmcpack inputs, both relevant
H5 data  are identical. Qmcpack will only read the determinant section and
basis set. Tested it by interchanging H5 files and passed my test. Of
course you will need to rename the H5 file.


On Jan 3, 2018 23:50, "Ye Luo" <notifications@github.com> wrote:

Does name="Gaussian" affect code path or it is just a name?
We don't need to test all the three 16-1, 4-4, 1-16 MPI-OMP combinations.
If I use hdf5, do I have identical qmcpack input files for gamess, QP,
pyscf?

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub
<https://github.com/QMCPACK/qmcpack/pull/607#issuecomment-355204538>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AXl6w_79ZD6_Kv27bz-Da47QQnnG3vQZks5tHGaIgaJpZM4RKg3E>
.

jtkrogel commented on src/QMCTools/PyscfParser.cpp:
```diff
@@ -0,0 +1,202 @@
+
+PyscfParser::PyscfParser()
+{
+  basisName = "Gaussian-G2";
```
Why leave "Gaussian-G2" here instead of "Gaussian" like everywhere else?

prckent replied:
Per @anbenali , the basis set name is always available in pyscf, so will be replaced with the actual name.

anbenali replied:
Paul is right, but that tag in that file is mainly a debugging legacy as I was trying to force the code to go through a specific path and did not know where it was picking that data.  I removed the tag from there since it is never read anyways. 

@prckent  and @jtkrogel In the case of multiple basis sets, a single name is impossible. by default I am putting Gaussian (Gaussian-G2 makes no sens). I first wanted to name it "HomeBrew" but I think apple has a trademark on it.. This does not do anything specific except allowing us to know what basis set is used.. any better ideas?

prckent commented:
(Removed the [WIP] since we are reviewing for merge. The work is mostly done.)

anbenali commented:
@prckent 
>>    1. Convert4qmc
Fixed. It was a bad if logic.

>> 2.    Remove: tests/converter/test_LiH_QP/LiH.ezfio.tar.gz
    
Done: QP run scripts added everywhere QP was used 

 >> 3.   Suffix inconsistency:
Done. 
 >> 4.    Add a test for the QP HDF5 output code path?
Was not necessary since the path os common to all format and as long as it works for one ot works for the others, but just for the kick of it, I added one anyways... 

>>5    Add a pyscf and qp converter and vmc test for FeCO6? Something sensitive to the d layout?
That was the worst example... FeCO6 is very difficult to converge... However, I still managed to have it. I uploaded the same example with gamess and Pyscf. However, the energies from the 2 codes are slightly different to begin with...  I found multiple cases while testing pyscf where I do not get the same energies as gamess or quantum package... but this has nothing to do with us... I will follow up with the pyscf people when I am not in 2D anymore... 

Last modification: 
PYSCF in Spherical representation is bugged in QMCPACK for larger than P shell. I do not have the courage to fix it for AOS. I will fix for SOA. I added an abort message if you try to run with spherical in pyscf and QMCPACK  will abort with a nice message.

ye-luo commented:
My concern is still on big files.
lines filename
 6,208 tests/converter/test_HDF5_LiH_qp/LiH.dump 
 10,317 tests/converter/test_HDF5_LiH_qp/gold.orbs.h5dump 
 6,208 tests/converter/test_LiH_qp/LiH.dump 
 7,214 tests/converter/test_LiH_qp/cc-pv5z 
 11,984 tests/converter/test_LiH_qp/gold.wfnoj.xml 
 13,380 tests/molecules/FeCO6_b3lyp_gms/C.BFD.xml 
 10,030 tests/molecules/FeCO6_b3lyp_gms/Fe.BFD.xml 
 13,380 tests/molecules/FeCO6_b3lyp_gms/O.BFD.xml 
 44,739 tests/molecules/FeCO6_b3lyp_gms/gamess/FeCO6.out 
 13,380 tests/molecules/FeCO6_b3lyp_pyscf/C.BFD.xml 
 10,030 tests/molecules/FeCO6_b3lyp_pyscf/Fe.BFD.xml 
 13,380 tests/molecules/FeCO6_b3lyp_pyscf/O.BFD.xml 
 5,991 tests/molecules/FeCO6_b3lyp_pyscf/pyscf/FeCO6.out 
 6,208 tests/molecules/LiH_dimer_ae_qp/quantum_package/LiH.dump 

LiH.dump appears three times. Probably keep the last one and replace all others with softlink.
h5dump should be all removed when using h5diff to replace all the h5dump (not by this PR).
FeCO6 has been in develop already, so all the BFD.xml should use softline to the existing ones.

anbenali commented:
@ye-luo 
Soft links to all existing files. 
However, I have no solution for codes outputs..... Any suggestion is welcome...

prckent commented:
Suggestion: require h5ls or h5diff to be available and working for these tests. Then we only need the gold h5 files and the dumps can be removed.

prckent commented:
I notice that the QP runs have a fairly large production level basis set. For testing one could use a much smaller basis. This would reduce the QP.dump files, although I don't think it is critical.

anbenali commented:
I agree... I do not really see the issue... It is not like we will want to "debug" those files... Worst case we can also just tar them if the line count is scary...

As for the h5ls/h5dump, this requires some cmake magic that I do not have. Somebody (@markdewing ?) can maybe answer/fix this issue? 

At this point, my second PR is almost ready to be pushed.. (SOA All electron with H5). However, it relies on this PR ;)

ye-luo commented:
If your second PR relies on this PR, then you can't submit it until this one is merged. Otherwise, you will see all the changes in PR repeated in the second PR. On the other hand, I thought the changes to enable hdf5 with SoA is pretty orthogonal to this PR and it can be separated from this PR.

ye-luo commented:
```
yeluo@ryzen-box:~/test$ git clone https://github.com/QMCPACK/qmcpack.git
Cloning into 'qmcpack'...
remote: Counting objects: 73792, done.
remote: Compressing objects: 100% (75/75), done.
remote: Total 73792 (delta 89), reused 135 (delta 82), pack-reused 73633
Receiving objects: 100% (73792/73792), 143.32 MiB | 10.46 MiB/s, done.
Resolving deltas: 100% (56900/56900), done.
Checking connectivity... done.
yeluo@ryzen-box:~/test$ git clone https://github.com/anbenali/qmcpack.git abenali
Cloning into 'abenali'...
remote: Counting objects: 74103, done.
remote: Compressing objects: 100% (119/119), done.
remote: Total 74103 (delta 121), reused 191 (delta 103), pack-reused 73874
Receiving objects: 100% (74103/74103), 149.40 MiB | 10.47 MiB/s, done.
Resolving deltas: 100% (57047/57047), done.
Checking connectivity... done.
```
The repo size will be increased when we merge this PR. 6MB is not much but also not small.
I'm wondering if we need to squash these intermediate commit which introduced and removed files.

The overall diskspace usage will increase from 490 MB to 508 MB.

anbenali commented:
I could indeed separate the 2,  but I will need to make sure I do not lose
anything in the process.

On Jan 11, 2018 12:52 AM, "Ye Luo" <notifications@github.com> wrote:

> yeluo@ryzen-box:~/test$ git clone https://github.com/QMCPACK/qmcpack.git
> Cloning into 'qmcpack'...
> remote: Counting objects: 73792, done.
> remote: Compressing objects: 100% (75/75), done.
> remote: Total 73792 (delta 89), reused 135 (delta 82), pack-reused 73633
> Receiving objects: 100% (73792/73792), 143.32 MiB | 10.46 MiB/s, done.
> Resolving deltas: 100% (56900/56900), done.
> Checking connectivity... done.
> yeluo@ryzen-box:~/test$ git clone https://github.com/anbenali/qmcpack.git abenali
> Cloning into 'abenali'...
> remote: Counting objects: 74103, done.
> remote: Compressing objects: 100% (119/119), done.
> remote: Total 74103 (delta 121), reused 191 (delta 103), pack-reused 73874
> Receiving objects: 100% (74103/74103), 149.40 MiB | 10.47 MiB/s, done.
> Resolving deltas: 100% (57047/57047), done.
> Checking connectivity... done.
>
> The repo size will be increased when we merge this PR. 6MB is not much but
> also not small.
> I'm wondering if we need to squash these intermediate commit which
> introduced and removed files.
>
> The overall diskspace usage will increase from 490 MB to 508 MB.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/QMCPACK/qmcpack/pull/607#issuecomment-356842867>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AXl6w5T_NTL2yh6iGwx-SO9QEZQvJZCHks5tJa-xgaJpZM4RKg3E>
> .
>

prckent commented:
@ye-luo , @markdewing If you would like a change here, please recommend a path forward. I should rerun the converters as a check, but Anouar has answered all my questions.

ye-luo commented:
I will look at the SoA HDF5 changes and see if that can be separated from this PR. If that is possible, I will look back to this PR and squash out those intermediate big file commits.

grahamlopez commented:
@ye-luo note that squashing out the commits won't remove the big files from the history/repo. I don't remember the best way off the top of my head, but it is a little bit more complicated to actually remove the files/data.

ye-luo commented:
@grahamlopez I think you are right that squashing doesn't delete the old commit from repo. But I think it is for anouar's fork, when a PR is merged, qmcpack repo only takes the commits from this PR but not everything from the other repo.

markdewing commented:
I can look into the CMake incantations to find h5diff and make sure it is available for improved gold file comparisons.   This change could be made in the future.

One addition (that can be made in a future PR) - the directions on using PyscfToQmcpack should be added to the PyscfToQmcpack.py file.

grahamlopez commented:
@ye-luo a few things: 1) squashing only "removes commits" by merging them together. All of the changes and files (i.e. the data) from the squashed commits remain. 2) during rebase when you would do a squash, you can also "drop" commits, which I _think_ should actually delete the data out of the branch. 3) when a PR is merged, it is merging a branch from the fork into this repository. You are right that not all data from the fork repo is brought over, but all data from the PR's source branch in the fork repo_is_ brought over.

ye-luo commented:
@grahamlopez Some commits introduced big files and then removed in a following commit. If I squashed them, the big files will be gone in the merged commit. Deleting commits is more risky I think, I may miss something. Once the history is rewritten, the big files no more exist in the branch history and will not land on qmcpack repo.

grahamlopez commented:
@ye-luo The problem is, that I suspect even if the files are no longer in the history, git still keeps the data around (for recovery purposes). Once a file has been in a history, it is non-trivial to get the actual data out of the repo.

grahamlopez commented:
This is what I'm talking about: https://help.github.com/articles/removing-files-from-a-repository-s-history/

ye-luo commented:
@prckent I have rewritten the history to fuse the commits with many or large files.

anbenali commented:
You are completely right.

The reason why I need to keep this file is simple. Because we put ourselves
in a relation of dependence to Gamess, we cannot use l>H... Since we chose
to put everything hard coded for cartesian representation following the
order in Gamess, and since Gamess does not support L>=5 and Li in cc-pv5z
has an H, and someone put the first test for LiH in cc-pv5z.... we need to
carry the truncated basis set...

If you notice in the pyscf input, I had also to hard code the basis set.

Now for this issue, if we just cut the basis set file and leave the data
for H and Li, it does the trick too.


On Fri, Jan 12, 2018 at 11:39 AM, Paul R. C. Kent <notifications@github.com>
wrote:

> Question: tests/converter/test_LiH_qp/cc-pv5z appears to be the entire
> basis set, i.e. has every element. Is this needed? Does it not come with
> QP? We could cut also down to H, Li only.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/QMCPACK/qmcpack/pull/607#issuecomment-357304489>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AXl6w5N9m0wUztU9AtZwrp8CbKIyzeakks5tJ5jVgaJpZM4RKg3E>
> .
>



-- 
-----------------------
Anouar Benali, PhD
Leadership Computing Facility
Argonne National Laboratory
Building 240 Office - 2127
9700 S Cass Av., Argonne Il, 60439
(630) 252-0058
